src/aws_xray_sdk_clj/impl.clj

```diff
@@ -95,69 +101,84 @@
+          (send-trace! (assoc arg-map :eid eid)))))
+    (if (= eid root-eid)
       (.endSegment recorder)
       (.endSubsegment recorder))))
 
-(defrecord AEntity [eid conn ^Clock clock ^AWSXRayRecorder recorder]
+(defrecord AEntity [^Clock clock conn done eid queue]
   ;; Implement AutoCloseable by default so that this can work with 'open-with'
   java.lang.AutoCloseable
   (close [entity]
     (protocol/-close! entity))
 
   protocol/IAutoCloseable
-  (-close! [{:as entity :keys [eid conn]}]
+  (-close! [{:keys [eid queue]}]
     (let [current-timestamp (current-timestamp-seconds clock)]
-      (d/transact! conn [{:db/id eid :end-at current-timestamp}]))
-    (when (= 1 eid)
-      (send-trace! entity))
-    nil)
+      (a/offer! queue [{:db/id eid :end-at current-timestamp}]))
+    (when (= eid root-eid)
+      (a/close! queue)))
 
   protocol/IEntity
-  (-set-exception! [{:as entity :keys [eid conn]} ex]
-    (d/transact! conn [{:db/id eid :exception ex}])
+  (-set-exception! [{:as entity :keys [eid queue]} ex]
+    (a/offer! queue [{:db/id eid :exception ex}])
     entity)
 
-  (-set-annotation! [{:as entity :keys [eid conn]} m]
+  (-set-annotation! [{:as entity :keys [eid queue]} m]
     (let [annotations (sanitize-keys m)]
-      (d/transact! conn [{:db/id eid :annotations annotations}])
+      (a/offer! queue [{:db/id eid :annotations annotations}])
       entity))
 
-  (-set-metadata! [{:as entity :keys [eid conn]} m]
+  (-set-metadata! [{:as entity :keys [eid queue]} m]
     (let [metadata (sanitize-keys m)]
-      (d/transact! conn [{:db/id eid :metadata metadata}])
+      (a/offer! queue [{:db/id eid :metadata metadata}])
       entity)))
 
 (extend-protocol protocol/IEntityProvider
   AWSXRayRecorder
   (-start! [^AWSXRayRecorder recorder {:as arg-map
-                                       :keys [clock]
-                                       :or {clock default-clock}}]
+                                       :keys [clock throttle queue]
+                                       :or {clock default-clock
+                                            throttle default-throttle
+                                            queue (a/chan (a/dropping-buffer 1000))}}]
     (let [conn (d/create-conn schema)
           current-timestamp (current-timestamp-seconds clock)
+          eid root-eid
           segment (merge (select-keys arg-map [:name :trace-id :parent-id])
-                         {:db/id 1 :start-at current-timestamp})]
-      (d/transact! conn [segment])
-      (->AEntity 1 conn clock recorder)))
+                         {:db/id eid :start-at current-timestamp})
+          ;; async worker
+          done (a/go-loop []
+                 (if-let [tx-data (<! queue)]
+                   (do
+                     (d/transact! conn tx-data)
+                     (recur))
+                   ;; try send the root segment if possible
+                   (try
+                     (dh/with-bulkhead throttle
```

Comment:
Is it correct to interpret that all send-trace! calls will be globally throttled unless a different throttle instance is passed in?

Reply:
Yes.

ref: https://failsafe.dev/bulkhead/

💡 I might be useful to dynamically determine which throttle to use based on the segment data. *i.e.* a throttle provider function.